### PR TITLE
RFC: Externals generation

### PR DESCRIPTION
**Pull request by diddledan**
I've been working on generating externals using this utility, and have worked-up the changes in this PR. Could you take a look to see how sane my changes are and comment on improvements I can make to get this into a mergable state?

There is an example of the output of `Run.hx` when parsing WordPress at https://gist.github.com/diddledan/b8d3907788052deac00c94a028f507df

I am very much novice with Haxe so I might have made terrible decisions :-)

### CODE REVIEW

File: peg/php/PType.hx [hunk 1]
```diff
@@ -4,10 +4,13 @@ enum PType {
 	TInt;
 	TFloat;
 	TString;
+	TArrayOfString;
```
Comment (RealyUniqueName) [inline peg/php/PType.hx]: `TArray` should be a generic type. `TArray(type:PType)`

File: peg/Parser.hx [hunk 2]
```diff
@@ -207,17 +211,30 @@ class Parser {
 		return name;
 	}
 
+	function mapParsedType(type:String):PType {
+		return switch(type) {
+			case 'int' | 'Int' | 'integer' | 'Integer': TInt;
+			case 'float' | 'Float': TFloat;
+			case 'string' | 'String': TString;
+			case 'string[]' | 'String[]': TArrayOfString;
```
Comment (RealyUniqueName) [inline peg/Parser.hx]: These and similar (e.g. `int[]`) should be parsed as `TArray(TString)`

File: peg/Parser.hx [hunk 3]
```diff
@@ -207,17 +211,30 @@ class Parser {
 		return name;
 	}
 
+	function mapParsedType(type:String):PType {
+		return switch(type) {
```
Comment (RealyUniqueName) [inline peg/Parser.hx]: instead of listing all possible casing in cases it's better to switch on `type.toLowerCase()`

File: peg/Parser.hx [hunk 4]
```diff
@@ -318,6 +335,11 @@ class Parser {
 
 		parseArguments(ctx, fn);
 
+		var rt = parseDocTagReturnType(fn);
+		if (rt != null && fn.returnType == TMixed) {
```
Comment (RealyUniqueName) [inline peg/Parser.hx]: makes sense to check `fn.returnType` before `parseDocTagReturnType(fn)` to avoid unnecessary parsing.

File: peg/Parser.hx [hunk 5]
```diff
@@ -337,6 +359,45 @@ class Parser {
 		return fn;
 	}
 
+	function parseDocTagReturnType(fn:PFunction):Null<PType> {
+		var documentedType:Null<String> = null;
+
+		if (docTagReturnTypeRE.match(fn.doc)) {
+			var type = docTagReturnTypeRE.matched(1);
+			if (combinationTypeRE.match(type)) {
+				type = 'mixed';
```
Comment (RealyUniqueName) [inline peg/Parser.hx]: Please, add a `//TODO` to not forget to implement correct parsing of `type1|type2`

File: peg/Parser.hx [hunk 6]
```diff
@@ -337,6 +359,45 @@ class Parser {
...
+	}
+
+	function parseDocTagParamType(fn:PFunction, paramName:String):Null<PType> {
+		var documentedType:Null<String> = null;
+
+		for (line in fn.doc.split('\n')) {
+			if (docTagParamsRE.match(line)) {
+				var type = docTagParamsRE.matched(1);
+				if (combinationTypeRE.match(type)) {
+					type = 'mixed';
```
Comment (RealyUniqueName) [inline peg/Parser.hx]: Please, add a `//TODO` to not forget to implement correct parsing of `type1|type2`

File: peg/Parser.hx [hunk 7]
```diff
@@ -352,7 +413,12 @@ class Parser {
 					return;
 				//$argName
 				case T_VARIABLE:
-					fn.addArg(parseVar(ctx, token.value));
+					var v = parseVar(ctx, token.value);
+					var t = parseDocTagParamType(fn, v.name);
+					if (t != null && v.type == TMixed) {
```
Comment (RealyUniqueName) [inline peg/Parser.hx]: Better check `v.type` before executing `parseDocTagParamType`

File: peg/Parser.hx [hunk 8]
```diff
@@ -369,6 +435,10 @@ class Parser {
 						case _: throw new UnexpectedTokenException(token);
 					}
 					v.type = type;
+					var t = parseDocTagParamType(fn, v.name);
+					if (t != null && v.type == TMixed) {
```
Comment (RealyUniqueName) [inline peg/Parser.hx]: Better check `v.type` before executing `parseDocTagParamType`

File: Run.hx [hunk 9]
```diff
@@ -1,13 +1,244 @@
...
+				'Array<Dynamic>';
+			case 'TObject':
+				'Map<String,Dynamic>';
+			case 'TCallable':
+				'Dynamic';
+			case 'TMixed':
+				'Dynamic';
+			case 'TResource':
+				'php.Resource';
+			case 'TClass':
```
Comment (RealyUniqueName) [inline Run.hx]:
```haxe
case TClass(c):
  switch (~/^...
```

File: Run.hx [hunk 10]
```diff
@@ -1,13 +1,244 @@
...
+					// These are in the haxe standard library
+					case 'Mysqli' | 'Mysqli_driver' | 'Mysqli_result' | 'Mysqli_stmt' |
+					'Mysqli_warning' | 'PDO' | 'PDOException' | 'PDOStatement' | 'SQLite3' |
+					'SQLite3Result' | 'SQLite3Stmt':
+						'php.db.${c}';
+					// These are in the haxe standard library
+					case 'ReflectionClass' | 'ReflectionFunctionAbstract' | 'ReflectionMethod' |
+					'ReflectionProperty' | 'Reflector':
+						'php.reflection.${c}';
+					// These require phpnatives (https://lib.haxe.org/p/phpnatives/)
```
Comment (RealyUniqueName) [inline Run.hx]: Unfortunately phpnatives groups php toplevel types into custom packages. That doesn't follow current scheme in haxe std lib for php.
It's ok to rely on phpnatives for now, but in future this have to be changed.
No need to change anything now, just making a note.

File: Run.hx [hunk 11]
```diff
@@ -1,13 +1,244 @@
...
+					case 'DOMAttr' | 'DOMCdataSection' | 'DOMCharacterData' | 'DOMComment' |
+					'DOMDocument' | 'DOMDocumentFragment' | 'DOMDocumentType' | 'DOMElement' |
+					'DOMEntity' | 'DOMEntityReference' | 'DOMException' | 'DOMImplementation' |
+					'DOMNamedNodeMap' | 'DOMNode' | 'DOMNodeList' | 'DOMNotation' |
+					'DOMProcessingInstruction' | 'DOMText' | 'DOMXPath' | 'LibXMLError' |
+					'SimpleXMLElement' | 'SimpleXMLIterator' | 'XMLReader' | 'XSLTProcessor':
+						'php.xml.${c}';
+					// All other/unknown classes
+					case _:
+						'${basePackage}.${c}';
```
Comment (RealyUniqueName) [inline Run.hx]: Should check if `basePackage` is empty. Otherwise the result could be `.MyClass` which is invalid.

File: Run.hx [hunk 12]
```diff
@@ -1,13 +1,244 @@
...
+				}
+			case _:
+				t.getName();
```
Comment (RealyUniqueName) [inline Run.hx]: Add a warning message here to easily locate missing types.

File: Run.hx [hunk 13]
```diff
@@ -1,13 +1,244 @@
...
+		};
+	}
+
+	static function getConst(c:peg.php.PConst):String {
+		return '${c.visibility} const ${c.name}:${getType(c.type)}';
```
Comment (RealyUniqueName) [inline Run.hx]: Php class constants are declared like this in externs:
```haxe
@:phpClassConst static public final name:Type;
```

Comment (RealyUniqueName) [reply Run.hx]: Oh, it's just for logging. Nevermind then.

File: Run.hx [hunk 14]
```diff
@@ -1,13 +1,244 @@
...
+		return '${v.visibility} var ${getVarName(v.name)}:${getType(v.type)}';
+	}
+
+	static function getFunction(fn:peg.php.PFunction, isNamespaceGlobal:Bool = false, ?namespace:String):String {
+		var args = fn.args.map(arg -> '${getVarName(arg.name)}:${getType(arg.type)}').join(', ');
+		var inlineCallParams = fn.args.map(arg -> getVarName(arg.name)).join(', ');
+		var callSite = '';
+		if (isNamespaceGlobal) {
+			var ns = namespace != '' ? '\\\\${namespace}\\\\': '';
+			callSite = ' return untyped __call__(\'${ns}${fn.name}\'${inlineCallParams != '' ? ', ' + inlineCallParams : ''})';
```
Comment (RealyUniqueName) [inline Run.hx]: use `php.Syntax.call` instead of `untyped __call__`

**Comment by RealyUniqueName** [comment]
Thanks for your contribution! It's a big step forward for PHP extern generator.

**Comment by diddledan** [comment]
Thank you for the thorough review :-) I'll get right onto the suggested changes!

File: Run.hx (hunk 14 above)
Comment (diddledan) [reply Run.hx]: done in 09f1f4c

File: Run.hx (hunk 13 above)
Comment (diddledan) [reply Run.hx]: done in 09f1f4c

File: Run.hx (hunk 12 above)
Comment (diddledan) [reply Run.hx]: done in 09f1f4c

File: Run.hx (hunk 11 above)
Comment (diddledan) [reply Run.hx]: done in 09f1f4c

File: Run.hx (hunk 9 above)
Comment (diddledan) [reply Run.hx]: done in 09f1f4c

File: Run.hx [hunk 15]
```diff
@@ -1,13 +1,244 @@
-import peg.*;
+import peg.PegException;
 
 class Run {
+	static var basePackage = '';
+
+	static function getType(t:peg.php.PType):String {
+		return switch t.getName() {
```
[the remark this answers is not in the document]
Comment (diddledan) [reply Run.hx]: done in 09f1f4c

File: peg/Parser.hx (hunk 8 above)
Comment (diddledan) [reply peg/Parser.hx]: done in 09f1f4c

File: peg/Parser.hx (hunk 7 above)
Comment (diddledan) [reply peg/Parser.hx]: done in 09f1f4c

File: peg/Parser.hx (hunk 6 above)
Comment (diddledan) [reply peg/Parser.hx]: done in 09f1f4c

File: peg/Parser.hx (hunk 5 above)
Comment (diddledan) [reply peg/Parser.hx]: done in 09f1f4c

File: peg/Parser.hx (hunk 4 above)
Comment (diddledan) [reply peg/Parser.hx]: done in 09f1f4c

File: peg/Parser.hx (hunk 3 above)
Comment (diddledan) [reply peg/Parser.hx]: done in 09f1f4c

File: peg/Parser.hx (hunk 2 above)
Comment (diddledan) [reply peg/Parser.hx]: done in 09f1f4c - I've implemented a recursion to handle this.

File: peg/php/PType.hx (hunk 1 above)
Comment (diddledan) [reply peg/php/PType.hx]: done in 09f1f4c - I've implemented a recursion to handle this.

**Comment by diddledan** [comment]
There's a bug that I'm not sure how to handle: https://github.com/RealyUniqueName/peg/pull/2/files#diff-143d53e0b6f3e16e8dc9a05eec9c3070R228 is using the value from `type.toLowerCase()` so for example `WP_Post` is getting saved into `TClass` as `wp_post`.. Would it be appropriate to swap:

```haxe
case path: TClass(path);
```

with:

```haxe
case _: TClass(type);
```

?

**Comment by diddledan** [comment]
I discovered I had missed-off parsing of doctags for `var` and `const` so I've added this. I also took the time to refactor the three doctag parsing functions which were largely similar into three `inline` functions which simply call the main parsing function with a dependency-injected regex specific to the item-type (variable, const, or parameter) that we want to match. This allows for less code duplication.

File: peg/Parser.hx [hunk 16]
```diff
@@ -337,6 +375,98 @@ class Parser {
...
+		var type = 'array';
+
+		var idx = 0;
+		if (line.indexOf('array') != -1) {
+			idx = line.indexOf('array') + 5;
+		} else if (line.indexOf('object') != -1) {
+			idx = line.indexOf('object') + 6;
+		}
+
+		for (chr in idx...line.length) {
```
Comment (RealyUniqueName) [inline peg/Parser.hx]: It's better for performance to iterate string and switch on chars like this:
```haxe
using StringTools; //add this before "class Parser {"
//<...>
for(chr => code in line) {
  switch code {
    case '<'.code:
      //<...>
```

Comment (RealyUniqueName) [reply peg/Parser.hx]: oh, wait. You're starting from `idx`, not from zero index. Nvm then

**Comment by RealyUniqueName** [comment]
Thanks!

**Pull request merged**